packages/flame/lib/src/game/transform2d.dart

```diff
@@ -0,0 +1,203 @@
+import 'dart:math' as math;
+import 'package:flutter/foundation.dart';
+import 'package:vector_math/vector_math_64.dart';
+import 'notifying_vector2.dart';
+
+/// This class describes a generic 2D transform, which is a combination of
+/// translations, rotations, reflections and scaling. These transforms are
+/// combined into a single matrix, that can be either applied to a canvas,
+/// composed with another transform, or used directly to convert coordinates.
+///
+/// The transform can be visualized as 2 reference frames: a "global" and
+/// a "local". At first, these two reference frames coincide. Then, the
+/// following sequence of transforms is applied:
+///   - translation to point [position];
+///   - rotation by [angle] radians clockwise;
+///   - scaling in X and Y directions by [scale] factors;
+///   - final translation by [offset], in local coordinates.
+///
+/// The class is optimized for repeated use: the transform matrix is cached
+/// and then recalculated only when the underlying properties change. Moreover,
+/// recalculation of the transform is postponed until the matrix is actually
+/// requested by the user. Thus, modifying multiple properties at once does
+/// not incur the penalty of unnecessary recalculations.
+///
+/// This class implements the [ChangeNotifier] API, allowing you to subscribe
+/// for notifications whenever the transform matrix changes. In addition, you
+/// can subscribe to get notified when individual components of the transform
+/// change: [position], [scale], and [offset] (but not [angle]).
+///
+class Transform2D extends ChangeNotifier {
+  final Matrix4 _transformMatrix;
+  bool _recalculate;
+  double _angle;
+  final NotifyingVector2 _position;
+  final NotifyingVector2 _scale;
+  final NotifyingVector2 _offset;
+  static const tau = 2 * math.pi;
+
+  Transform2D()
+      : _transformMatrix = Matrix4.identity(),
+        _recalculate = true,
+        _angle = 0,
+        _position = NotifyingVector2(),
+        _scale = NotifyingVector2()..setValues(1, 1),
+        _offset = NotifyingVector2() {
+    _position.addListener(_markAsModified);
+    _scale.addListener(_markAsModified);
+    _offset.addListener(_markAsModified);
+  }
+
+  factory Transform2D.copy(Transform2D other) => Transform2D()
+    ..angle = other.angle
+    ..position = other.position
+    ..scale = other.scale
+    ..offset = other.offset;
+
+  /// Check whether this transform is equal to other, up to the given
+  /// [tolerance]. Setting tolerance to zero will check for exact equality.
+  /// Transforms are considered equal if their rotation angles are the same
+  /// or differ by a multiple of 2π, and if all other transform parameters:
+  /// translation, scale, and offset are the same.
+  ///
+  /// The [tolerance] parameter is in absolute units, not relative.
+  bool closeTo(Transform2D other, {double tolerance = 1e-10}) {
```

Comment:
I don't think it matters, but the tolerance value here would be a much larger change for for example `scale` than for `position`, in most cases.

Reply:
Yes, that's the downside of absolute error... I mostly used it because it's slightly less code to write, but perhaps it's better to switch to relative tolerance then?

Reply:
I don't think it will make much difference in reality since epsilon probably will be quite small in most cases anyways, I think we can just re-visit this if we experience problems with it in the future.